File: benchmark/nodejs/duffs_device.js
function duffsDevice(count) {
    let a = new Array(count).fill(0);
    let b = new Array(count).fill(1);
    
    let n = Math.floor(count / 8);
    let rem = count % 8;
    let i = 0;
    
    while (rem > 0) {
        a[i] = b[i];
        i++;
        rem--;
    }
    
    while (n > 0) {
        a[i] = b[i]; i++;
        a[i] = b[i]; i++;
        a[i] = b[i]; i++;
        a[i] = b[i]; i++;
        a[i] = b[i]; i++;
        a[i] = b[i]; i++;
        a[i] = b[i]; i++;
        a[i] = b[i]; i++;
        n--;
    }
    return a[count - 1];
}

let start = Date.now();
let res = 0;
for (let k = 0; k < 100; k++) {
    res = duffsDevice(100000);
}
let end = Date.now();

console.log("Result: " + res);
console.log("Time: " + Math.floor(end - start) + "ms");

File: benchmark/lua/duffs_device.lua
function duffs_device(count)
    local a, b = {}, {}
    for i = 1, count do
        a[i] = 0
        b[i] = 1
    end
    
    local n = math.floor(count / 8)
    local rem = count % 8
    local i = 1
    
    while rem > 0 do
        a[i] = b[i]
        i = i + 1
        rem = rem - 1
    end
    
    while n > 0 do
        a[i] = b[i]; i = i + 1
        a[i] = b[i]; i = i + 1
        a[i] = b[i]; i = i + 1
        a[i] = b[i]; i = i + 1
        a[i] = b[i]; i = i + 1
        a[i] = b[i]; i = i + 1
        a[i] = b[i]; i = i + 1
        a[i] = b[i]; i = i + 1
        n = n - 1
    end
    return a[count]
end

local start = os.clock() * 1000
local res = 0
for k = 1, 100 do
    res = duffs_device(100000)
end
local end_time = os.clock() * 1000

print("Result: " .. res)
print("Time: " .. math.floor(end_time - start) .. "ms")
